src/hotspot/share/opto/opaquenode.hpp

```diff
@@ -101,17 +101,29 @@ class OpaqueZeroTripGuardNode : public Opaque1Node {
 class OpaqueMultiversioningNode : public Opaque1Node {
 private:
   bool _is_delayed_slow_loop;
+  bool _is_useful;
 
 public:
   OpaqueMultiversioningNode(Compile* C, Node* n) :
-      Opaque1Node(C, n), _is_delayed_slow_loop(true)
+      Opaque1Node(C, n), _is_delayed_slow_loop(true), _is_useful(true)
   {
     init_class_id(Class_OpaqueMultiversioning);
   }
   virtual int Opcode() const;
   virtual const Type* bottom_type() const { return TypeInt::BOOL; }
   bool is_delayed_slow_loop() const { return _is_delayed_slow_loop; }
-  void notify_slow_loop_that_it_can_resume_optimizations() { _is_delayed_slow_loop = false; }
+
+  void notify_slow_loop_that_it_can_resume_optimizations() {
+    assert(_is_useful, "must still be useful");
+    _is_delayed_slow_loop = false;
+  }
+
+  void set_useless() {
```

Comment:
I suggest `mark_useless()` to be in line with `ParsePredicateNode`.
```suggestion
  void mark_useless() {
```